### PR TITLE
fix(udfs): return NULL for a vector distance that is not defined (refs #11263)

### PR DESCRIPTION
**Pull request by claudespice**
## Summary

A vector carrying `NaN` or an infinity has no cosine distance, L2 distance, inner product or norm — but the kernels answered one anyway, and the two dispatch paths did not agree on which one.

| input | `FixedSizeList<Float32>` (SIMD) | `List` / `LargeList` (scalar) |
|---|---|---|
| element is `NaN` / `±Inf` | **0** — distance 0, an exact match | 0.5 |
| zero-magnitude vector | 0.5 | 0.5 |

The `FixedSizeList` column is what embedding columns actually use, and `vector_search` passes its query vector as a `FixedSizeList` literal, so the production path is the one that answered **"perfect match"**. `_score` is `1 - distance`, so a failed embedding scored `1.0` and sorted above every real result; under `EmbeddingAggregation::Max` one corrupt chunk took its whole document to the top.

Root cause is that the guard on that path screened the kernel's **output** for non-finiteness. `simsimd`'s `cos` normalizes and returns a finite `0` for a `NaN` row, and nothing about that output distinguishes it from two genuinely parallel vectors — so an output guard cannot catch it.

## Changes

- Screen the **input** at the shared chokepoint, not the output. Only `Cosine` needs it: `Dot` and `L2Squared` accumulate their inputs directly, so a non-finite element reaches their output and the existing check catches it for free. `Kernel::hides_non_finite_input` records which is which, and the shared test drives all three kernels, so if that ever stops holding the test fails rather than a guard silently going missing.
- The scalar cosine kernels screen their three accumulators, which covers every non-finite element without a per-element branch. This check must precede the zero-magnitude branch — a `NaN` vector measured against a *zero* vector would otherwise be answered 0.5 on the strength of the zero side alone. That ordering has its own regression test.
- `l2_norm` needs no input screen at all: a sum of squares has no cancelling terms, so a non-finite element always reaches the result.
- Zero-magnitude vectors are unchanged. They are defined input with no direction, and both paths still call them orthogonal (0.5).

NULL is already a reachable score today (a null embedding row produces one), and the score aggregates are `max`/`avg`/`sum`, which skip nulls — so a poisoned chunk stops competing rather than starting to.

## Performance impact

This is a hot path, so it was measured, not assumed — `cargo bench -p runtime-datafusion-udfs --bench cosine_distance`, 10000 rows x 1536 dims:

| | trunk | this PR |
|---|---|---|
| cosine (SIMD) | 4.72 ms | 7.65 ms (**+62%**) |
| `inner_product`, `l2_distance`, `l2_norm` | — | unaffected (free output check) |

The screen's formulation was tuned against that benchmark; four alternatives measured worse and are recorded in the code so they are not retried: short-circuiting `iter().all(f32::is_finite)` (+296%, blocks auto-vectorization), a float `acc += v * 0.0` chain (+600%, serial FP dependency), an 8-lane manual split (+110%), and scanning both whole buffers up front (+60%).

Roughly **half** the remaining cost is rescanning the broadcast query literal once per row; removing it needs a signature change to the shared `make_scalar_function`, so it is filed as #13090 rather than grown into this PR. Correctness is the priority here per CLAUDE.md, but the number is real and a maintainer should see it.

## Not fixed here

- **The formula divergence in the same issue** — local `(1 - sim) / 2` vs federated `1 - sim`. It changes user-visible `_score` on *every* row and drifts the whole search snapshot suite, and #11263 explicitly parks it for a maintainer decision. This PR changes only which rows are NULL, not the scale of any defined score.
- #13088 — the federated path unparses to `array_cosine_distance` and does not inherit this NULL contract.
- #13089 — the write-side guards test `all(v == 0.0 || v.is_nan())` instead of `any(!v.is_finite())`, and DuckDB's own `validate_vector` is never called on write.

## Test plan
- [x] Added regression test for a non-finite element on every dispatch path (`FixedSizeList`, `List`, `LargeList`, `Float64` elements) and every kernel (`Dot`, `L2Squared`, `Cosine`, `l2_norm`)
- [x] Added edge case tests for non-finite-vs-zero-magnitude in both argument orders, zero-magnitude staying orthogonal, and a clean row in the same batch as a poisoned one still being scored
- [x] Each of the four guards was independently neutered and confirmed to fail a test, so none of them is vacuous — the neuter is also what surfaced the missing non-finite-vs-zero-vector case
- [x] `make lint` passes (workspace-wide `cargo fmt --all -- --check` + full clippy — not a per-crate scoped invocation)
- [x] `make test` passes
- [x] `make signoff` run after the final push (`Attestation` check is green)

## Review gate

- Adversarial review: **skipped** — `codex` exited 1: "Your workspace is out of credits"; `grok` not installed. Its angles were run by hand instead, and two produced changes: the FSL path was verified to be what `vector_search` actually uses (`embeddings/udtf.rs:885`), and the performance question was measured rather than asserted — which is what caught a 4x regression in the first formulation.
- `/security-review`: no findings. The diff adds no SQL, I/O, auth, or logging; the behavior it removes was itself the exploitable direction (a writable `NaN` embedding ranked first in every search, a ranking-manipulation primitive in a RAG pipeline).
- `/simplify`: applied — dropped a redundant whole-buffer prescan in `l2_norm`, inlined a wrapper that had become a pass-through, replaced hand-rolled `Option` matches with `append_option`, and replaced two duplicated test builders with one generic helper in the shared `testing` module. Three findings were out of scope and filed as #13088, #13089 and #13090. Light checks re-run green.

Refs #11263 — deliberately **not** `Fixes`. The issue's headline defect is the local-vs-federated
formula divergence listed under *Not fixed here*, which #11263 itself parks for a maintainer
decision; this PR fixes the non-finite-input hazard filed alongside it. #11263 stays open until
that decision is made.

### CODE REVIEW

**Comment by claudespice** [comment]
@copilot review

File: crates/runtime-datafusion-udfs/src/cosine_distance.rs [hunk 1]
```diff
@@ -181,7 +191,9 @@ pub(crate) fn cosine_distance_inner(args: &[ArrayRef]) -> DataFusionResult<Array
         // divide by 2 to map into the standard [0, 1] distance range.
         // Zero-magnitude vectors have an undefined direction; simsimd treats the
         // dot product as zero (orthogonal), yielding distance 0.5.
-        (FixedSizeList(_, _), FixedSizeList(_, _)) => compute_fsl_f32_cosine(args),
+        (FixedSizeList(_, _), FixedSizeList(_, _)) => {
+            compute_fsl_f32(args, Kernel::Cosine, |v| v / 2.0)
```
[the remark this answers is not in the document]
Comment (claudespice) [reply crates/runtime-datafusion-udfs/src/cosine_distance.rs:195]: You're right, and this was the one thing in the PR that could do real damage on merge — thanks for catching it.

The body already had a *Not fixed here* section saying the `(1 - sim) / 2` vs `1 - sim` divergence is out of scope, and #11263 itself says *"Not fixing unilaterally because either direction changes user-visible score values; flagging for a maintainer decision."* Closing the issue anyway would have retired that decision without anyone making it — and the divergence is the issue's headline, not a footnote.

Fixed by retargeting the linkage rather than by widening the PR, since the formula choice is explicitly a maintainer call:

- Title: `(fixes #11263)` → `(refs #11263)` — the squash-merge commit inherits the PR title, so the keyword had to come off both places, not just the body.
- Body trailer: `Fixes #11263` → `Refs #11263`, with the reason stated inline so a future reader doesn't "helpfully" restore the closing keyword.

#11263 stays open for the formula decision. This PR keeps its actual scope: which rows are NULL, not the scale of any defined score.

File: crates/runtime-datafusion-udfs/src/vector_simd.rs [hunk 2]
```diff
@@ -235,12 +276,28 @@ where
                 continue;
             }
         }
+        // A non-finite element makes the whole row's score undefined, and the
+        // kernels do not report that themselves: `simsimd`'s `cos` answers 0 —
+        // distance 0, an exact match — for a row carrying NaN, which puts a
+        // failed embedding at the top of `ORDER BY _score DESC`. Screen the
+        // input rather than the output, because a fabricated score is finite.
+        if screen_input && (!all_finite(slice_a) || !all_finite(slice_b)) {
```
Comment (bjchambers) [inline crates/runtime-datafusion-udfs/src/vector_simd.rs:284]: Do we need the screen when simsimd is available? IIUC, we're screening and making the result *match* what simsimd would return. In that case, it seems like we could make the screening conditional on *not* using simsimd, which would also preserve the performance of the path when simsimd is available.

Comment (claudespice) [reply crates/runtime-datafusion-udfs/src/vector_simd.rs:284]: The screen is needed *because* simsimd is the kernel — it's the simsimd path that hides the defect, so conditioning it on "not simsimd" would drop the guard exactly where the bug lives.

`SpatialSimilarity::cos` answers **0** for a row carrying `NaN`. Cosine distance 0 is an exact match, so `_score = 1 - distance` came out `1.0` and a failed embedding sorted above every real result; under `EmbeddingAggregation::Max` one corrupt chunk took its whole document to the top. Nothing about that output distinguishes it from two genuinely parallel vectors, which is why the output check can't catch it and the input has to be screened. So the screen isn't making the result match what simsimd returns — it's making it NULL, which is what neither path returns on its own.

There's also no non-simsimd path here to condition on. `simsimd = "6.5"` is an unconditional dependency (no feature gate), `Kernel::apply` always dispatches to it, and simsimd selects AVX-512 / AVX2 / NEON / scalar internally at load time — this code has no visibility into which. The other dispatch path, `List` / `LargeList`, never reaches this function: it screens its three accumulators in `cosine_distance.rs:473`, which is a separate and much cheaper guard.

The screen *is* already conditional, on the kernel rather than on simsimd — `Kernel::hides_non_finite_input()` is true only for `Cosine`. `Dot` and `L2Squared` accumulate their inputs directly, so a non-finite element reaches their output and the existing output check catches it for free; they skip the per-element scan entirely and pay nothing. `l2_norm` likewise.

On the cost, which is measured rather than assumed: **+62%** on the cosine SIMD path (4.72 ms -> 7.65 ms at 10000 rows x 1536 dims). Four cheaper formulations measured worse and are recorded in the code so they don't get retried. Roughly half of what remains is rescanning the broadcast query literal once per row — a real win that costs no correctness, needing a signature change to the shared `make_scalar_function`, which is why it's filed as #13090 rather than grown into this PR.

Resolving on that basis, but reopen if you read the tradeoff differently — the +62% is the one number here a maintainer might weigh against the ranking bug.

**Review by lukekim — APPROVED** [review]
